taggify.js

```diff
@@ -0,0 +1,696 @@
+      resize_input(this);
+      renderSuggestions(this, this.parentElement.children[3]);
+    });
+
+    input.parentElement.addEventListener('focusout', function(e) {
+      if (!document.hasFocus(this.children[3])) {
+        clearSuggestions(this.children[3]);
+      }
+    });
+  }
+
+  window.Taggify['taggify'] = taggify;
+
+  function printWidths(widths) {
+    if (!SETTINGS.debugging) {
+      return;
+    }
+    var todelete = document.getElementsByClassName('delme');
+    for (var i = 0; i < todelete.length; i++) {
+      todelete[i].parentElement.removeChild(todelete[i]);
+    }
+    for (var i = 0; i < widths.length+2; i = i+2) {
+      var div = document.createElement('div');
+      if (i % 4 == 0) {
+        div.style.background = 'black';
+      } else {
+        div.style.background = 'white';
+      }
+      div.style.height = '10px';
+      div.style.position = 'absolute';
+      div.style.left = widths[i];
+      div.style.top = '5em';
+      div.className = 'delme';
+      div.style.width = widths[i+1] - widths[i];
+      document.body.appendChild(div);
+    }
+  }
+
+  function resize_input(inpt, min_width = 10) {
+    var style = "font-family:"+css(inpt, 'font-family')+";font-size:"+css(inpt, 'font-size')+";";
+    var width = getTextWidth(inpt.value, style)+inpt.value.length*2+4;
+    if (width < min_width) { width = min_width; }
+    inpt.style.width = width+"px";
+  }
+
+  function generateSuggestionBox(name, desc, input = '') {
+    var ele = document.createElement('div');
+    ele.setAttribute('data-tag', name);
+    ele.className = 'suggestion';
+    ele.setAttribute('tabindex', 0);
+    var header_container = document.createElement('h3');
+    var header = document.createElement('span');
+    if (input.length > 0) {
+      name = htmlEncode(name).replace(htmlEncode(input), "<b><u>"+htmlEncode(input)+"</u></b>");
+      header.innerHTML = name;
+    } else {
+      header.textContent = name;
+    }
+    header.className = 'tag';
+    header_container.appendChild(header);
+    var body = document.createElement('p');
+    body.textContent = desc;
+    ele.appendChild(header_container);
+    ele.appendChild(body);
+    return ele;
+  }
+
+  function moveCaretToEnd(input) {
+    setTimeout(function(){ input.selectionStart = input.selectionEnd = 10000; }, 0);
+  }
+
+  function clearSuggestions(suggestionEle) {
+    while (suggestionEle.firstChild) {
+      suggestionEle.removeChild(suggestionEle.lastChild);
+    }
+  }
+
+  function renderSuggestions(input, suggestionEle) {
+    clearSuggestions(suggestionEle);
+    var suggestions = Taggify.autocomplete_dict;
+    console.log(suggestions);
+    for (var i = 0; i < suggestions.length; i++) {
+      if (suggestions[i][0].includes(input.value)) {
+        var name = suggestions[i][0];
+        var desc = suggestions[i][1];
+        suggestionEle.appendChild(generateSuggestionBox(name, desc, input.value));
+      }
+    }
+  }
+}(window, document));
```

Comment:
You have an undefined parameter (which we've had words about already...), but you don't assign anything to it - that makes it an _undefined_ undefined, which is _really_ hard to define. (Honestly, I'd just remove that parameter.)